kratos/utilities/voxel_utilities.h

```diff
@@ -0,0 +1,173 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ariadna Cortes
+//
+//
+
+#if !defined(KRATOS_VOXEL_UTILITIES)
+#define  KRATOS_VOXEL_UTILITIES
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/geometrical_object.h"
+#include "includes/node.h"
+#include "geometries/geometry.h"
+#include "geometries/hexahedra_3d_8.h"
+#include "intersection_utilities.h"
+
+namespace Kratos
+{ 
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class VoxelUtilities
+ * @ingroup KratosCore
+ * @brief Utilities to compute the real volume inside a voxel
+ * @details This class provides static methods to compute (using different approximations) the portion of a 
+ * voxel that is actually filled with volume, according to the known triangle elements that intersect the voxel
+ * @author Ariadna Cortes
+ */
+class VoxelUtilities
+{
+public:
+
+    ///@name Type Definitions
+    ///@{
+
+    typedef Node<3> NodeType;
+    typedef Node<3>::Pointer NodePtrType;
+    typedef Geometry<NodeType> GeometryType;
+    typedef GeometryType::Pointer GeometryPtrType;
+    typedef GeometryType::GeometriesArrayType GeometryArrayType;
+    typedef GeometryType::PointsArrayType PointsArrayType;
+
+    /// Pointer definition of VoxelInsideVolume
+    KRATOS_CLASS_POINTER_DEFINITION( VoxelUtilities );
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    /**
+     * @brief Default constructor
+     */
+    VoxelUtilities(){}
+
+    /// Destructor
+    virtual ~VoxelUtilities(){}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Aproximates the actual volume inside the voxel 
```

Comment:
Looking into your code, you are calculating the volume fraction rather than the volume. And you are not mentioning that the volume fraction is for the nodes with negative distance